core/src/main/java/com/linecorp/armeria/internal/client/DefaultClientRequestContext.java

```diff
@@ -531,6 +528,14 @@ private DefaultClientRequestContext(DefaultClientRequestContext ctx,
 
         this.endpointGroup = endpointGroup;
         updateEndpoint(endpoint);
+        if (ctx.endpoint() == endpoint || endpoint == null) {
+            eventLoop = ctx.eventLoop().withoutContext();
+        } else {
+            acquireEventLoop(endpoint);
+        }
```

Comment:
> When creating a derived client in RetryingClient, a new endpoint is selected for each retry, but since the EventLoop of the parent is used as is. That causes the Endpoint can't use the existing connection pool for multiplexing and makes a new connection.

Maybe it's just me but I didn't understand this change, sorry 😅 

Looking at the change:
- If endpoint == parent.endpoint, then the same eventloop is used
- If endpoint != parent.endpoint, then a new eventloop is assigned

Even if the same eventloop is used, wouldn't the same connection be used anyways?

The new logic tries to acquire a new eventloop if `parent.eventloop != eventloop`. How does this avoid creating a new connection?

Reply:
Good question. Let me illustrate how we can get a pooled channel.
- `EventLoop` is used as a key to get an `HttpChannelPool` which manages the lifecycle of `PooledChannel`s.
https://github.com/line/armeria/blob/dc2bdc6107bb1c6784773781cac60cac6ca39203/core/src/main/java/com/linecorp/armeria/client/HttpClientDelegate.java#L172
- `DefaultEventLoopScheduler` memorizes an acquired `EventLoop` to an `EventLoop` internally for at least 1 minute.
https://github.com/line/armeria/blob/bb791585e0787f4ded8fb91796911b728552aba1/core/src/main/java/com/linecorp/armeria/client/DefaultEventLoopScheduler.java#L176

If we assign the parent event loop directly to the client context, the mapping between `EventLoop` and `HttpChannelPool` in `DefaultEventLoopScheduler` is ignored and `factory.pool(ctx.eventLoop().withoutContext())` may not contain the `Pooled`Channel` for the child `Endpoint`.

For that reason, `acquireEventLoop(endpoint)` is used to get the pooled connection.

Reply:
Thanks for the explanation 👍